ocaml/xapi/xapi_psr.ml

```diff
@@ -145,11 +167,214 @@ functor
+
+    let tell_cleanup_old_pool_secret host =
+      Helpers.call_api_functions ~__context (fun rpc session_id ->
+          Client.Host.cleanup_pool_secret rpc session_id host)
+
+    let cleanup_master =
+      cleanup_internal
+        ~additional_files_to_remove:[checkpoint_path; checkpoint_tmp_path]
   end
 
-let start ~__context = failwith "not impl"
+let notify_new ~new_ps =
+  match !Xapi_globs.pool_secrets with
+  | _ :: _ :: _ ->
+      D.info "xapi_psr.ml:notify_new: already accepting new_ps"
+  | _ ->
+      let old_ps = Xapi_globs.pool_secret () in
+      SecretString.write_to_file old_pool_secret_backup_path old_ps ;
+      SecretString.write_to_file new_pool_secret_backup_path new_ps ;
+      Xapi_globs.pool_secrets := [old_ps; new_ps]
+
+let notify_send ~new_ps =
+  if SecretString.equal (Xapi_globs.pool_secret ()) new_ps then
+    D.info "xapi_psr.ml:notify_send: already sending new_ps"
+  else
+    match !Xapi_globs.pool_secrets with
+    | [] | [_] ->
+        raise
+          Api_errors.(
+            Server_error
+              (internal_error, ["pool secret rotation not in progress"]))
+    | old_pool_secret :: new_pool_secret :: _ ->
+        let () =
+          Unixext.with_file new_pool_secret_backup_path [Unix.O_RDONLY] perm
+            (fun ifd ->
+              Unixext.with_file !Xapi_globs.pool_secret_path
+                [Unix.O_WRONLY] perm (fun ofd ->
+                  try
+                    let (_ : int64) = Unixext.copy_file ifd ofd in
+                    ()
+                  with e ->
+                    D.error
+                      "xapi_psr.ml:notify_send: copy pool_secret_path failed" ;
+                    raise e))
+        in
+        Xapi_globs.pool_secrets := [new_pool_secret; old_pool_secret] ;
+        Db_globs.pool_secret :=
+          new_pool_secret |> SecretString.rpc_of_t |> Db_secret_string.t_of_rpc
+
+let cleanup = cleanup_internal ~additional_files_to_remove:[]
+
+module HostSet = Set.Make (struct
+  type t = API.ref_host
+
+  let compare = Stdlib.compare
+end)
+
+module Ptoken = Genptokenlib.Lib
+
+(* at most one PSR should be
+   running at any given time *)
+let m = Mutex.create ()
+
+let start ~__context =
+  (* checked preconditions:
+     * we are the master
+     * HA disabled
+     * RPU not running
+     * all hosts in pool 'alive' *)
+  let f () =
+    if
+      not
+        (Helpers.is_pool_master ~__context
+           ~host:(Helpers.get_localhost ~__context))
+    then
+      raise
+        Api_errors.(
+          Server_error (host_is_slave, [Pool_role.get_master_address ()])) ;
+    let live_hosts = Helpers.get_live_hosts ~__context |> HostSet.of_list in
+    let all_hosts_list = Xapi_pool_helpers.get_master_slaves_list ~__context in
+    let all_hosts = all_hosts_list |> HostSet.of_list in
+    ( if not (HostSet.equal all_hosts live_hosts) then
+        let offline_host =
+          HostSet.diff all_hosts live_hosts
+          |> HostSet.find_first (fun _ -> true)
```

Comment:
Is this the same as `HostSet.min_elt`? You could also start with the set difference and save the `equal`.